components/script/dom/eventtarget.rs

```diff
@@ -187,22 +187,21 @@ impl CompiledEventListener {
 
                     CommonEventHandler::BeforeUnloadEventHandler(ref handler) => {
                         if let Some(event) = event.downcast::<BeforeUnloadEvent>() {
-                            let rv = event.ReturnValue();
-
+                            // Step 5
                             if let Ok(value) = handler.Call_(object,
                                                              event.upcast::<Event>(),
                                                              exception_handle) {
-                                match value {
-                                    Some(value) => {
-                                        if rv.is_empty() {
-                                            event.SetReturnValue(value);
-                                        }
-                                    }
-                                    None => {
-                                        event.upcast::<Event>().PreventDefault();
+                                let rv = event.ReturnValue();
```

Comment:
https://html.spec.whatwg.org/multipage/webappapis.html#the-event-handler-processing-algorithm Step 5 says we should be setting `E`'s cancelled flag which I believe is done via `mark_as_handled()`.

Reply:
Thanks for looking deeper in that part. It appers there are some things that need clarifying, and after looking more into it as well, here are a few points I think we could consider:

1. Step 5 of "the event handler processing algorithm" writes "Set E's canceled flag", which itself [is defined](https://dom.spec.whatwg.org/#canceled-flag) as a "flag" so either set or unset, and should be set when `preventDefault` is called.
2. In servo, `canceled` [can be either](https://github.com/servo/servo/blob/4f4fb32264ab94406c189471e9ec753a06011b23/components/script/dom/event.rs#L367) `Allowed`, `Prevented`, or `Handled`.
3. calling `PreventDefault` [sets canceled to](https://github.com/servo/servo/blob/4f4fb32264ab94406c189471e9ec753a06011b23/components/script/dom/event.rs#L253) `Prevented`. 
4. `mark_as_handled`  [sets canceled to](https://github.com/servo/servo/blob/4f4fb32264ab94406c189471e9ec753a06011b23/components/script/dom/event.rs#L205) `Handled`. 
5. Step 7 of https://html.spec.whatwg.org/multipage/browsing-the-web.html#prompt-to-unload-a-document and Step 9 of https://html.spec.whatwg.org/multipage/browsing-the-web.html#unload-a-document both write "If any event listeners were triggered", without further defining this concept. 

From the above, I deduct that:

A. Step 5 of "#the-event-handler-processing-algorithm" would in servo translate to "setting `canceled` to `EventDefault::Prevented`(or calling `Event.PreventDefault`).
 I actually missed one part of that step, see the additional commit(to be squashed). 
B. "If any event listeners were triggered" would in servo translate to `canceled` being either `EventDefault::Prevented` or `EventDefault::Handled`, yet it's unclear where that should be done.

I feel like the spec is actually unclear in this regard and could be improved:
1. "If any event listeners were triggered" seems to only relate to the "salvageable" state of the document in the context of (before-)unloading, yet it relates to both "beforeunload" and "unload" events(so it cannot be dealt with only at the level of BeforeUnloadEvent).
2. The spec is missing a general concept of 'the event was handled'.
3. Such a concept could be introduced either as part of [#concept-event-listener-inner-invoke](https://dom.spec.whatwg.org/#concept-event-listener-inner-invoke), which is the implementation I initially went for, or as part of [#the-event-handler-processing-algorithm](https://html.spec.whatwg.org/multipage/webappapis.html#the-event-handler-processing-algorithm)
4. While in servo we could implement this by setting `canceled` to `EventDefault::Handled`, when it wasn't previously set to `EventDefault::Prevented`, in the spec it would probably be clearer to do it as a separate `handled` flag, that could be set at either the end of "#the-event-handler-processing-algorithm" or anytime after Step 1 of "#concept-event-listener-inner-invoke"

What do you think?

Reply:
I see what you mean with `EventDefault::Handled` vs `EventDefault::Prevented`. It looks like `EventDefault::Handled` is only used when some input events. Not sure why we have `Allowed` `Prevented` and `Handled`, but yes in this case we should just call `PreventDefault` here.

Reply:
Do you mean "mark_as_handled"? 😀 
So we actually need this `Handled`,  to set the document's salvageable state to false when there was any listener triggered for `beforeunload` or `unload`. And I've tried calling `PreventDefault`, and unfortunately that makes a bunch of tests fail that make assertions about wether the default was prevented or not. 

So in another words, we need to mark the event as 'handled', but not 'prevented', hence my somewhat hacky code that you've highligthed. 

Lastly, we need this for the `unload` event as well as `beforeunload`, so it cannot be done only inside `CommonEventHandler::BeforeUnloadEventHandler`(by the way does cover Step 5 "Otherwise", [here](https://github.com/servo/servo/blob/4f4fb32264ab94406c189471e9ec753a06011b23/components/script/dom/eventtarget.rs#L218))

Reply:
asked for input via https://github.com/whatwg/html/issues/3646

Reply:
Ok good news, this has been resolved I think see https://github.com/whatwg/html/issues/3646#issuecomment-384581127
I'll add a commit to deal with this...